core/lib/zksync_core/src/state_keeper/state_keeper_storage.rs

```diff
@@ -0,0 +1,243 @@
+        conn: &mut StorageProcessor<'_>,
+        rocksdb: RocksDB<StateKeeperColumnFamily>,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> anyhow::Result<Option<PgOrRocksdbStorage<'a>>> {
+        tracing::debug!("Catching up RocksDB synchronously");
+        let rocksdb_builder = RocksdbStorage::builder(rocksdb)
+            .await
+            .context("Failed initializing RocksDB storage")?;
+        let rocksdb = rocksdb_builder
+            .synchronize(conn, stop_receiver)
+            .await
+            .context("Failed to catch up state keeper RocksDB storage to Postgres")?;
+        if let Some(rocksdb) = &rocksdb {
+            let rocksdb_l1_batch_number = rocksdb.l1_batch_number().await.unwrap_or_default();
+            tracing::debug!(%rocksdb_l1_batch_number, "Using RocksDB-based storage");
+        } else {
+            tracing::warn!("Interrupted")
+        }
+        Ok(rocksdb.map(|rocksdb| rocksdb.into()))
+    }
+
+    async fn access_storage_inner<'a>(
+        &'a self,
+        rt_handle: Handle,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> Option<PgOrRocksdbStorage<'a>> {
+        match self {
+            AsyncRocksdbCache::Postgres(pool) => Some(
+                Self::access_storage_pg(rt_handle, pool)
+                    .await
+                    .expect("Failed accessing Postgres storage"),
+            ),
+            AsyncRocksdbCache::Rocksdb(rocksdb, pool) => {
+                let mut conn = pool
+                    .access_storage_tagged("state_keeper")
+                    .await
+                    .expect("Failed getting a Postgres connection");
+                Self::access_storage_rocksdb(&mut conn, rocksdb.clone(), stop_receiver)
+                    .await
+                    .expect("Failed accessing RocksDB storage")
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl ReadStorageFactory for AsyncRocksdbCache {
+    type ReadStorageImpl<'a> = PgOrRocksdbStorage<'a>;
+
+    async fn access_storage<'a>(
+        &'a self,
+        rt_handle: Handle,
+        stop_receiver: &watch::Receiver<bool>,
+    ) -> Option<Self::ReadStorageImpl<'a>> {
+        self.access_storage_inner(rt_handle, stop_receiver).await
+    }
+}
+
+impl<T: ReadStorageFactory> StateKeeperStorage<T> {
+    pub fn new(inner: Arc<Mutex<T>>) -> StateKeeperStorage<T> {
+        Self { inner }
+    }
+
+    pub fn factory(&self) -> T {
+        // It's important that we don't hold the lock for long; that's why `T` implements `Clone`
+        self.inner.lock().expect("poisoned").clone()
+    }
+}
+
+impl StateKeeperStorage<AsyncRocksdbCache> {
+    pub fn async_rocksdb_cache(
+        pool: ConnectionPool,
+        state_keeper_db_path: String,
+        enum_index_migration_chunk_size: usize,
+    ) -> Self {
+        let inner = Arc::new(Mutex::new(AsyncRocksdbCache::Postgres(pool.clone())));
+        let factory = inner.clone();
+        tokio::task::spawn(async move {
```

Comment:
Please define a type for the task so that it can be used in the [node framework](https://github.com/matter-labs/zksync-era/tree/main/core/node/node_framework) that @popzxc is developing. Also, in any case, you want to make this task cancelable (i.e., having it take `stop_receiver` rather than creating a surrogate one), and to propagate errors for it (so that if an errors occurs in it, it can be monitored and leads to the node termination). Maybe, we'd even want to have a [healthcheck](https://github.com/matter-labs/zksync-era/tree/main/core/lib/health_check) for this task to monitor syncing progress (naturally, this can be done in a follow-up PR). You can look at the API server as an example of the component with multiple tasks.